### ISSUE TITLE
support cartesian product of infinite iterators

### ISSUE THREAD

**Issue by smichr** [sympy/sympy#17157]
Danny Yoo (anyone know where he is?) has a [recipe](http://code.activestate.com/recipes/52271-how-to-take-the-cross-product-of-two-infinite-sequ/?in=user-98032) that produces the cartesian product (cross product) of infinite iterators. `itertools.product` does not appear to support this. Perhaps the utilities.cartes could be made to recognize when an infinite sequence/iterator is included and adapt accordingly.

**Comment by oscarbenjamin** [comment]
I used to see Danny Yoo on the Python tutor mailing list. Haven't seen him there for some time though...

In general there is no way to know that an iterator is infinite without iterating it (I guess this is the halting problem...). It might be useful in any case to have an icartes that can handle infinite iterators.

**Comment by cag** [comment]
I wrote it in JS if you wanna port it to Python:

```js
function* product(...arrs) {
  if (arrs.length === 0) yield [];
  else
    for (const h of arrs[0])
      for (const ts of product(...arrs.slice(1))) yield [h, ...ts];
}

function* xproduct(...iters) {
    const encountered = iters.map(() => [])
    let done = false
    while(!done) {
        done = true
        for(let i = 0; i < iters.length; i++) {
            if(!iters[i].done) {
                done = false
                const newValue = iters[i].next().value
                encountered[i].push(newValue)

                yield* product(...encountered.map((e, j) => i === j ? [newValue] : e))
            }
        }
    }
}
```

It needs to remember what it has seen already though in order to construct the new tuples.

**Comment by oscarbenjamin** [comment]
I think we do need this since this hangs:
```julia
In [2]: iter(ProductSet(Integers))                                                                                                
^C---------------------------------------------------------------------------
KeyboardInterrupt                         Traceback (most recent call last)
<ipython-input-2-d9571a34a744> in <module>
----> 1 iter(ProductSet(Integers))

~/current/sympy/sympy/sympy/sets/sets.py in __iter__(self)
    787         """
    788         if self.is_iterable:
--> 789             return product(*self.sets)
    790         else:
    791             raise TypeError("Not all constituent sets are iterable")

~/current/sympy/sympy/sympy/sets/fancysets.py in __iter__(self)
    191         while True:
    192             yield i
--> 193             yield -i
    194             i = i + 1
    195 
```
It seems itertools.product can't work with infinite iterators even in the trivial case: it tries to consume the entire iterator during its constructor.

This also means that iterable hangs:
```julia
In [2]: from sympy.core.compatibility import iterable                                                                             

In [3]: iterable(ProductSet(Integers))                                                                                            
^C---------------------------------------------------------------------------
KeyboardInterrupt
```

So just asking if an object is iterable can lead to an infinite loop even if we don't actually iterate over it.

**Comment by oscarbenjamin** [comment]
Danny Yoo's recipe isn't directly usable since it supposes indexible inputs (rather than just iterable inputs).

This is a simple product function that will return the expected results for finite iterators and won't hang for infinite ones:
```python
def iproduct_simple(*iterables):
    '''Returns the Cartesian product of iterables but doesn't hang for
    infinite iterators'''
    if len(iterables) == 0:
        yield ()
        return
    first, others = iterables[0], iterables[1:]
    for f in first:
        for o in iproduct_simple(*others):
            yield (f,) + o
```
The drawback is that if there are any infinite iterators iproduct_simple isn't guaranteed to yield any particular element.

This version should yield every element eventually:
```python
def iproduct2(iterable1, iterable2):
    '''Cartesian product of two possibly infinite iterables'''

    it1 = iter(iterable1)
    it2 = iter(iterable2)

    elems1 = []
    elems2 = []

    sentinel = object()
    def append(it, elems):
        e = next(it, sentinel)
        if e is not sentinel:
            elems.append(e)

    n = 0
    append(it1, elems1)
    append(it2, elems2)

    while n <= len(elems1) + len(elems2):
        for m in range(n-len(elems1)+1, len(elems2)):
            yield (elems1[n-m], elems2[m])
        n += 1
        append(it1, elems1)
        append(it2, elems2)

def iproduct(*iterables):
    '''Returns Cartesian product of iterables. Yields every element
    eventually'''
    if len(iterables) == 0:
        yield ()
        return
    elif len(iterables) == 1:
        for e in iterables[0]:
            yield (e,)
    elif len(iterables) == 2:
        for e12 in iproduct2(*iterables):
            yield e12
    else:
        first, others = iterables[0], iterables[1:]
        for ef, eo in iproduct2(first, iprod(*others)):
            yield (ef,) + eo
```

**Comment by oscarbenjamin** [comment]
I've included this in #17593
